File: problem72.clj
(defn sieve-multiple [sieve ^long current-index ^long until]
  (let [^double current-number (get sieve current-index)]
    (if (= current-number (double current-index))
      (reduce (fn [result index]
                (let [^double number (get sieve index)]
                  (assoc! sieve index 
                          ^double (- number (/ number current-number))))) 
              sieve (range current-index until current-index))
      sieve)))

(defn multiple-sieve [^long until]
  (let [sieve (range 2 (inc until))]
    (reduce + 0 (persistent!
                 (reduce (fn [result index]
                           (sieve-multiple result index (inc until)))
                         (transient (into [0.0 0.0] (map double sieve))) 
                         sieve)))))

(time (multiple-sieve 1000000))

File: problem72.hs
import Control.Monad
import Data.Array.IO
import Data.Array

collectMultiply :: Int -> IOUArray Int Int -> Int -> IO (IOUArray Int Int)
collectMultiply current array index = 
  do
    elem <- readArray array index
    writeArray array index (elem - (elem `quot` current))
    return array

-- given a number and an array, returns an array of the same size
-- with all the elements updated accordingly
sieve :: Int -> IOUArray Int Int -> Int -> IO (IOUArray Int Int)
sieve until array index =
  do
    current <- readArray array index
    if current /= index
      then return array
      else foldM (collectMultiply index) array [index, index+index .. until]
      
solve72 :: Int -> IO Int
solve72 until = do
  array <- (genArray until)
  solved <- foldM (sieve until) array [2 .. until] >>= getElems :: IO [Int]
  return $ sum solved

genArray :: Int -> IO (IOUArray Int Int)
genArray until = newListArray (2,until) [2 .. until]
